src/dbnode/storage/shard.go

```diff
@@ -2660,74 +2661,217 @@ func (s *dbShard) Repair(
-		if err == io.EOF {
-			break
+		for seriesIter.Next() {
+			frame := seriesIter.Current()
+			unit, singleUnit := frame.Units().SingleValue()
+			annotation, singleAnnotation := frame.Annotations().SingleValue()
+
+			if frameValues := frame.Values(); len(frameValues) > 0 {
+				lastIdx := len(frameValues) - 1
+
+				value := frameValues[lastIdx]
+				timestamp := frame.Timestamps()[lastIdx]
+
+				if !singleUnit {
+					// TODO: what happens if unit has changed mid-tile?
+					// Write early and then do the remaining values separately?
+					if lastIdx < len(frame.Units().Values()) {
+						unit = frame.Units().Values()[lastIdx]
+					}
+				}
+
+				if !singleAnnotation {
+					// TODO: what happens if annotation has changed mid-tile?
+					// Write early and then do the remaining values separately?
+					if lastIdx < len(frame.Annotations().Values()) {
+						annotation = frame.Annotations().Values()[lastIdx]
+					}
+				}
+
+				dp.Timestamp = timestamp
+				dp.Value = value
+				err = encoder.Encode(dp, unit, annotation)
+				if err != nil {
+					s.metrics.largeTilesWriteErrors.Inc(1)
+					multiErr = multiErr.Add(err)
+				}
+
+				processedTileCount.Inc()
+			}
 		}
-		if err != nil {
-			return processedBlockCount, err
+
+		segment := encoder.DiscardReset(opts.Start, segmentCapacity, targetSchemaDescr)
+
+		segmentLen := segment.Len()
+		if segmentLen > segmentCapacity {
+			// Will use the same capacity for the next series.
+			segmentCapacity = segmentLen
 		}
 
-		data.IncRef()
-		bytesReader.Reset(data.Bytes())
-		dataPointIter.Reset(bytesReader, nil)
+		writerData[0] = segment.Head.Bytes()
+		writerData[1] = segment.Tail.Bytes()
+		checksum := segment.CalculateChecksum()
 
-		for dataPointIter.Next() {
-			dp, unit, annot := dataPointIter.Current()
-			_, err = s.writeAndIndex(ctx, id, tags, dp.Timestamp, dp.Value, unit, annot, wOpts, true)
-			if err != nil {
-				s.metrics.largeTilesWriteErrors.Inc(1)
-				lastWriteError = err
-			}
+		if err := writer.WriteAll(id.Bytes(), encodedTags, writerData, checksum); err != nil {
+			s.metrics.largeTilesWriteErrors.Inc(1)
+			multiErr = multiErr.Add(err)
+		} else {
+			s.metrics.largeTilesWrites.Inc(1)
 		}
 
-		dataPointIter.Close()
+		id.Finalize()
+		//FIXME apparently this does not finalize segment.Tail as FinalizeTail flag is not set by default.
+		segment.Finalize()
+	}
 
-		data.DecRef()
-		data.Finalize()
+	if err := writer.Close(); err != nil {
+		multiErr = multiErr.Add(err)
+	} else {
+		// Notify all block leasers that a new volume for the namespace/shard/blockstart
+		// has been created. This will block until all leasers have relinquished their
+		// leases.
+		_, err = s.opts.BlockLeaseManager().UpdateOpenLeases(block.LeaseDescriptor{
+			Namespace:  s.namespace.ID(),
+			Shard:      s.ID(),
+			BlockStart: opts.Start,
+		}, block.LeaseState{Volume: nextVersion})
+	}
 
-		processedBlockCount++
+	if err := readerIter.Err(); err != nil {
```

Comment:
Hm I would check the readerIter.Err() first, if we failed to iterate through series altogether (i.e. immediately fail) we don't want to write out an empty new volume. 

So maybe
```
if err := readerIter.Close(); err != nil {
  multiErr = multiErr.Add(err)
  // Do not close the writer, since that will write out a checkpoint file. (maybe there should be some abort method to the FS streaming writer so it at least closes the FDs and deletes the files? probably ok to not delete the files though too since partial stuff should get cleaned up by cleanup manager)
} else {
  if err := writer.Close(); err != nil {
    multiErr = multiErr.Add(err)    
  } else {
    // Update block lease
  }
}
```

Reply:
Adding `StreamingWriter.Abort` method to skip a checkpoint file and close the file descriptors.

Reply:
Sounds good 👍